### ISSUE TITLE
Conv BatchNorm folding failed: steps towards reducing this error

### ISSUE THREAD

**Issue by sanderland** [intel/intel-extension-for-pytorch#250]
When calling optimize on a transformer model, you get a warning "Conv BatchNorm folding failed during the optimize process.".
I did some work in trying to trace where this comes from, and steps towards improving

## concrete_args
The first step where it fails is here:

```python
if input_ids is not None and inputs_embeds is not None:
    raise ValueError("You cannot specify both input_ids and inputs_embeds at the same time")
```
Where both are Proxy objects

This can be prevented by allowing concrete_args and pass it through to the inner function where it is used.
```
optimization.fuse(optimized_model, inplace=inplace, concrete_args=concrete_args)
```

Alternatively (or additionally) any parameter not in `sample_input` could be made a concrete_arg equal to it's default value.

## .dim() checks
Passing concrete_args makes it fail at
```python
if not (attention_mask.dim() == 2 and self.config.is_decoder):
    # show warning only if it won't be shown in `create_extended_attention_mask_for_decoder`
    if device is not None:
        warnings.warn(
            "The `device` argument is deprecated and will be removed in v5 of Transformers.", FutureWarning
        )
```
Although the above is of course not particularly relevant, there are other dim checks, and they are fairly common in code bases.

My though here was that sample_input could provide such a dimension, but I am not familiar enough with the proxy setup to know if and how this can be done.

**Comment by jgong5** [comment]
IPEX leverages PyTorch's FX function to fold the conv and bn while sometimes the folding could fail due to FX limitations like input-dependent conditionals as you observed. To workaround such a problem, you can identify some sub-modules that can be FX-traceable and call the `optimization.fuse` on them manually before calling `ipex.optimize`. May I know if you are working with vision transformer models that contain conv and BN in it?

**Comment by sanderland** [comment]
@jgong5 No, they are mainly NLP models which do not even contain this. However, if the same technique is used to fuse e.g. layernorm in the future, the same issue would arise. I was also curious if the FX function can work around the dim issue at all, or if this is a fundamental limitation.

**Comment by jgong5** [comment]
I don't think folding layernorm is valid since it computes mean/var on the fly. For FX on HF models, you can try out HF's FX tracing alternative: https://github.com/huggingface/transformers/blob/main/src/transformers/utils/fx.py

**Comment by sanderland** [comment]
@jgong5 huggingface uses ipex to optimize in a generic way, but ipex can essentially not handle any model (including resnet), and always throws a noisy and confusing warning. 
This issue is about discussing if ipex could be improved to make this step can be made not to fail all the time.
Alternatively, there could perhaps be a way to turn off these warnings, or bypass the step when it can not possibly succeed.

Your last suggestion confuses me, I do not see a way to pass such a custom tracing module to .optimize

**Comment by jgong5** [comment]
>  huggingface uses ipex to optimize in a generic way, but ipex can essentially not handle any model (including resnet), and always throws a noisy and confusing warning.

I'm fine to disable the bn folding here. I noticed you submitted a PR accordingly.

> Your last suggestion confuses me, I do not see a way to pass such a custom tracing module to .optimize

Oh, I'm talking about FX traceability for HF models in general, not something related to ipex.optimize.

**Comment by sanderland** [comment]
On trying some more things, the potential for improvements seem mostly on the PyTorch side. I still think the warning could be improved though, as it seems to make you think you had conv+bn to fuse, but fusing them failed, rather than "the model contains if statements".
